docs/shadertrap_language.md

```diff
@@ -0,0 +1,338 @@
+
+### DECLARE_SHADER
+
+```
+DECLARE_SHADER result KIND kind
+// GLSL code
+END
+```
+
+Declares a shader, which can subsequently be compiled. The shader will be validated using [`glslang`](https://github.com/KhronosGroup/glslang).
+
+- The declared shader is identified by `result`
+- `kind` specifies the kind of shader this is, and must be one of `VERTEX`, `FRAGMENT` or `COMPUTE`, for a vertex, fragment or compute shader, respectively. Other shader kinds are not supported.
+- The shader text is taken from the lines following `DECLARE_SHADER`, up to but not including a line starting with `END`
+
+The reason for separating the declaration and compilation of shaders into two separate commands is that it allows writing a test that compiles the same shader multiple times without having to duplicate the text of the shader.
+
+### DUMP_RENDERBUFFER
+
+```
+DUMP_RENDERBUFFER RENDERBUFFER renderbuffer FILE file
+```
+
+Dumps a renderbuffer to a file, in PNG format.
+
+- `renderbuffer` is the renderbuffer to be dumped, and must be produced by `CREATE_RENDERBUFFER`
+- `file` is the file to which the PNG will be written
+
+### RUN_COMPUTE
+
+```
+RUN_COMPUTE PROGRAM compute_program NUM_GROUPS x y z
+```
+
+- `compute_program` must be a *compute* program produced by `CREATE_PROGRAM`
+- `x`, `y` and `z` must be non-negative integers specifying the number of work groups in the *x*, *y* and *z* dimensions, respectively, that should execute the compute program
+
+### RUN_GRAPHICS
+
+```
+RUN_GRAPHICS
+  PROGRAM graphics_program
+  VERTEX_DATA
+    [ attribute_index_1 -> BUFFER vertex_buffer_1 OFFSET_BYTES offset_1 STRIDE_BYTES stride_1 DIMENSION dimension_1,
+      attribute_index_2 -> BUFFER vertex_buffer_2 OFFSET_BYTES offset_2 STRIDE_BYTES stride_2 DIMENSION dimension_2,
+      ...
+      attribute_index_n -> BUFFER vertex_buffer_n OFFSET_BYTES offset_n STRIDE_BYTES stride_n DIMENSION dimension_n,
+    ]
+  INDEX_DATA index_buffer
+  VERTEX_COUNT count
+  TOPOLOGY topology
+  FRAMEBUFFER_ATTACHMENTS
+    [ location_1 -> attachment_1,
+      location_2 -> attachment_2,
+      ...
+      location_n -> attachment_n
+    ]
+```
+
+Runs a graphics workload.
+
+- `graphics_program` must be a *graphics* program produced by `CREATE_PROGRAM`.
+- The `VERTEX_DATA` parameter is a mapping with one entry for each vertex attribute associated with the vertex shader of `graphics_program`. If the vertex shader declares an `in` variable with `layout(location = i)` then `VERTEX_DATA` must provide an entry of the form `i -> ...`.  In this entry:
+  - `vertex_buffer_i` specifies a buffer from which vertex data for vertex attribute `i` will be taken. This must be produced by `CREATE_BUFFER`. The elements of this buffer will be interpreted as 32-bit floating-point numbers - i.e. the `GL_FLOAT` type is used for vertex data.
```

Comment:
I know that we don't want to cover every possibility right now, but should we maybe think about types other than 32 bit floats, just to make sure we don't miss something? I guess it is not a big deal, as we can make breaking changes, if needed. I feel like, on mobile, using smaller data types is quite important.

Reply:
Good point - I've opened #30 to track it.